File: src/Binary_Tree_Right_Side_View.py
# -*- coding: utf-8 -*-

from Queue import Queue


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def bfs(self, root):
        q, ret = Queue(), []
        tmp, end = None, root
        q.put(root)
        while not q.empty():
            node = q.get()
            if node.left:
                q.put(node.left)
                tmp = node.left
            if node.right:
                q.put(node.right)
                tmp = node.right
            if node == end:
                end = tmp
                ret.append(node.val)
        return ret

    # @param {TreeNode} root
    # @return {integer[]}
    def rightSideView(self, root):
        if root is None:
            return []
        result = self.bfs(root)
        return result


if __name__ == "__main__":
    vals = [1, 2, 3, 4, 5]
    nodes = map(TreeNode, vals)
    nodes[0].left, nodes[0].right = nodes[1], nodes[2]
    nodes[1].right, nodes[2].right = nodes[4], nodes[3]
    solution = Solution()
    print solution.rightSideView(nodes[0])

File: src/algorithms/cpp/Binary_Tree_Right_Side_View.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
#include <queue>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        queue <TreeNode *> q;
        vector <int> result;
        if (root == NULL)       return result;
        TreeNode *end = root, *nxt_end = root;
        q.push(root);
        while (!q.empty()) {
            TreeNode *node = q.front();
            q.pop();
            if (node->left)     q.push(node->left), nxt_end = node->left;
            if (node->right)    q.push(node->right), nxt_end = node->right;
            if (node == end) {
                result.push_back(node->val);
                end = nxt_end;
            }
        }
        return result;
    }
};

int main() {
    return 0;
}
